pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,665 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// defaultMaxReplicaUpdateCount refers to the maximum number of creation or deletion of AzVolumeAttachment objects in a single ManageReplica call
+	defaultMaxReplicaUpdateCount = 1
+	updateAttemptInterval        = time.Duration(1) * time.Millisecond
+	updateTimeout                = time.Duration(1) * time.Minute
+)
+
+type AttachmentProvisioner interface {
+	PublishVolume(ctx context.Context, volumeID string, nodeID string, volumeContext map[string]string) (map[string]string, error)
+	UnpublishVolume(ctx context.Context, volumeID string, nodeID string) error
+}
+
+type ReconcileAttachDetach struct {
+	client                client.Client
+	azVolumeClient        azVolumeClientSet.Interface
+	kubeClient            kubeClientSet.Interface
+	namespace             string
+	attachmentProvisioner AttachmentProvisioner
+}
+
+var _ reconcile.Reconciler = &ReconcileAttachDetach{}
+
+func (r *ReconcileAttachDetach) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, request.Name, request.Namespace, true)
+	// if object is not found, it means the object has been deleted. Log the deletion and do not requeue
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// detachment request
+	if requestedToBeDeleted(azVolumeAttachment) {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.Attached || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed || azVolumeAttachment.Status.State == v1alpha1.DetachmentFailed {
+			if err := r.triggerDetach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// attachment request
+	} else if azVolumeAttachment.Status.Detail == nil {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed {
+			if err := r.triggerAttach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// promotion request
+	} else if azVolumeAttachment.Spec.RequestedRole != azVolumeAttachment.Status.Detail.Role {
+		if err := r.promote(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileAttachDetach) triggerAttach(ctx context.Context, attachmentName string, azVolumeAttachment *v1alpha1.AzVolumeAttachment) error {
+	// initialize metadata and update status block
+	conditionFunc := func() (bool, error) {
```

Comment:
Consider defining the condition function closer to the point of use, i.e. the `PollImmediate` call.